### PR TITLE
Expanding Bundle Segmentation Tutorial

### PR DESCRIPTION
**Pull request by BramshQamar**
In this PR, I plan to expand the Bundle Segmentation using RecoBundles tutorial. 

- [x] Explain what different parameters do
- [x] Add multiple RB outputs to show how changing these parameters will affect the output so users can make an informed decision
- [x] Add auto-calibration RB part in the example 
- [x] Change default parameters to higher values so it extracts denser bundles as most users prefer them
- [x] Corrected the orientation of bundles in the figures

Bramsh

### CODE REVIEW

File: doc/examples/bundle_extraction.py [hunk 1]
```diff
@@ -115,7 +115,28 @@
...
+Extracting bundles using RecoBundles [Garyfallidis17]_
+
+RecoBundles requires a model (reference) bundle and tries to extract similar
+looking bundle from the input tractogram. There are some key parameters that
+users can set as per requirements. Here are the key threshold parameters and
+their function in Recobundles:
+
+    - model_clust_thr : It will use QuickBundles to get the centroids of the
+    model bundle and work with centroids instead of all streamlines. This helps
+    in making RecoBundles faster. If you prefer to use all the streamnlines of
```
Comment (arokem) [inline doc/examples/bundle_extraction.py:127]: I would add 
```suggestion
    to make RecoBundles faster. If you prefer to use all the streamlines of
```

Comment (arokem) [inline doc/examples/bundle_extraction.py:127]: I would add here that the larger the number is the fewer centroids there are and also explain what the units are -- this is mm? Voxel? Some kind of probability? What is the range of acceptable values?

File: doc/examples/bundle_extraction.py [hunk 2]
```diff
@@ -115,7 +115,28 @@
...
+    the model bundle, you can set this threshold to 0.01.
+
+    - reduction_thr : This threshold will be used to reduce the search space
+    for finding the streamlines that match model bundle streamlines in shape.
+    Instead of looking at the entire tractogram, now we will be looking at
+    neighboring region of a model bundle in the tractogram. Increase the threshold
+    to increase the search space.
```
Comment (arokem) [inline doc/examples/bundle_extraction.py:134]: Again: what are the units? What's a reasonable range of values?

File: doc/examples/bundle_extraction.py [hunk 3]
```diff
@@ -115,7 +115,28 @@
...
+
+    - pruning_thr : This threshold will filter the streamnlines that have distance
```
Comment (arokem) [inline doc/examples/bundle_extraction.py:136]:
```suggestion
    - pruning_thr : This threshold will filter the streamlines that have distance
```

File: doc/examples/bundle_extraction.py [hunk 4]
```diff
@@ -115,7 +115,28 @@
...
+    between model bundle and them greater than the pruning_thr. This serves as
```
Comment (arokem) [inline doc/examples/bundle_extraction.py:137]: Maybe : "This threshold will filter the streamlines for which the distance to the model bundle is greater than ..."

File: doc/examples/bundle_extraction.py [hunk 5]
```diff
@@ -115,7 +115,28 @@
...
+    filtering the neighborhood area (search space) to get model bundle like
```
Comment (arokem) [inline doc/examples/bundle_extraction.py:138]:
```suggestion
    between model bundle and them greater than the pruning_thr. This serves to
    filter the neighborhood area (search space) to get streamlines that are like the model bundle.
```

**Review by arokem — COMMENTED** [review]
Thanks for adding this! A few small comments.

**Comment by BramshQamar** [comment]
Hi @skoudoro 

PR is ready for review!

File: doc/examples/bundle_shape_similarity.py [hunk 6]
```diff
@@ -56,13 +57,21 @@ def show_both_bundles(bundles, colors=None, show=True, fname=None):
         scene.add(streamtube_actor)
     if show:
         window.show(scene)
-    elif fname is not None:
-        window.record(scene, out_path=fname, size=(900, 900))
+    if fname is not None:
+        sleep(1)
```
Comment (skoudoro) [inline doc/examples/bundle_shape_similarity.py:61]: Why this sleep here?

**Review by skoudoro — APPROVED** [review]
Looks good to me. Can you confirm @arokem? if yes, can you merge this PR?

File: doc/examples/bundle_assignment_maps.py [hunk 7]
```diff
@@ -80,9 +80,9 @@
 
 scene = window.Scene()
 scene.SetBackground(1, 1, 1)
-scene.add(actor.streamtube(model_af_l, colors=disks_color))
+scene.add(actor.line(model_af_l, colors=disks_color))
```
Comment (arokem) [inline doc/examples/bundle_assignment_maps.py:83]: This looks a bit nicer:
```suggestion
scene.add(actor.line(model_af_l, colors=disks_color, fake_tube=True, linewidth=6))
```

File: doc/examples/bundle_assignment_maps.py [hunk 8]
```diff
@@ -41,9 +41,9 @@
 
 scene = window.Scene()
 scene.SetBackground(1, 1, 1)
-scene.add(actor.streamtube(model_af_l))
+scene.add(actor.line(model_af_l))
```
Comment (arokem) [inline doc/examples/bundle_assignment_maps.py:44]:
```suggestion
scene.add(actor.line(model_af_l, fake_tube=True, linewidth=6))
```

File: doc/examples/bundle_extraction.py [hunk 9]
```diff
@@ -153,70 +207,229 @@
...
+"""
+
+reco_af_l = StatefulTractogram(target[af_l_labels], target_header,
+                               Space.RASMM)
+save_trk(reco_af_l, "AF_L_org_3.trk", bbox_valid_check=False)
+
+
+"""
+Let's apply auto-calibrated RecoBundles on the output of standard RecoBundles.
+This step will filter out the fuzzy streamlines. This time, the RecoBundles'
```
Comment (arokem) [inline doc/examples/bundle_extraction.py:363]:
```suggestion
This step will filter out the fuzzy streamlines. This time, the RecoBundles'
```
What does "fuzzy" mean here? It looks like maybe it filters outliers. Is that what is meant here?

File: doc/examples/bundle_extraction.py [hunk 10]
```diff
@@ -153,70 +207,229 @@
...
+extracted bundle will serve as a model bundle. As a rule of thumb, provide
+larger threhsold values in standard RecoBundles function and smaller values in
```
Comment (arokem) [inline doc/examples/bundle_extraction.py:365]:
```suggestion
larger threshold values in standard RecoBundles function and smaller values in
```

**Review by arokem — COMMENTED** [review]
This is great! The additional details in the recobundles tutorial are _really_ helpful. 

I get this warning when I run that example, though:
```
/Users/arokem/miniconda3/envs/dipy/lib/python3.8/site-packages/dipy/segment/bundles.py:684: UserWarning: Streamlines do not have the same number of points. All streamlines need to have the same number of points. Use dipy.tracking.streamline.set_number_of_points to adjust your streamlines
```
Do we need to resample all the streamlines, before passing them to rb?

**Comment by BramshQamar** [comment]
> This is great! The additional details in the recobundles tutorial are _really_ helpful.
> 
> I get this warning when I run that example, though:
> 
> ```
> /Users/arokem/miniconda3/envs/dipy/lib/python3.8/site-packages/dipy/segment/bundles.py:684: UserWarning: Streamlines do not have the same number of points. All streamlines need to have the same number of points. Use dipy.tracking.streamline.set_number_of_points to adjust your streamlines
> ```
> 
> Do we need to resample all the streamlines, before passing them to rb?

This warning seems to be coming from `bundles_distances_mdf` function from `dipy.tracking.distances` file.

We calculate mdf distance in RecoBundles, and it needs streamlines to have same number of points. Let me fix this.

File: doc/examples/bundle_shape_similarity.py (hunk 6 above)
Comment (BramshQamar) [reply doc/examples/bundle_shape_similarity.py]: removed

**Comment by skoudoro** [comment]
Also, this warning might have disappeared. it has been fixed in #2373.

Maybe you just need to rebase @BramshQamar. I believe you did not start this branch with the lastest repo version

**Comment by BramshQamar** [comment]
> Also, this warning might have disappeared. it has been fixed in #2373.
> 
> Maybe you just need to rebase @BramshQamar. I believe you did not start this branch with the lastest repo version

This PR changes error to a warning. I was thinking of working on a different PR to change the way we call mdf function in RecoBundles so that users won't get this warning.

**Comment by skoudoro** [comment]
Any additional comments here?  @arokem? @BramshQamar?

File: doc/examples/bundle_extraction.py [hunk 11]
```diff
@@ -107,43 +107,97 @@
...
+    Default is set to mdf.
+
+    - slr : If slr flag is set to True, local registration of model bundle with
+    neighbouring area will be performed. Default and recommended is True.
+"""
 
 """
-Extracting bundles using recobundles [Garyfallidis17]_
+Read Arcuate Fasciculus Left bundle from already fetched atlas data to use it
+as model bundle. Let's visualize the model bundle.'
```
Comment (arokem) [inline doc/examples/bundle_extraction.py:150]:
```suggestion
as model bundle. Let's visualize the model bundle.
```

**Review by arokem — COMMENTED** [review]
One more tiny thing, but after this I think it's ready to go!

**Comment by skoudoro** [comment]
Thank you @BramshQamar ! Looking forward to your next PR

**Pull request merged**